README.md
# GitHub Action: await-remote-run

[![GitHub Workflow Status](https://img.shields.io/github/actions/workflow/status/codex-/await-remote-run/test.yml?style=flat-square)](https://github.com/Codex-/await-remote-run/actions/workflows/test.yml) [![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier) [![codecov](https://img.shields.io/codecov/c/github/Codex-/await-remote-run?style=flat-square)](https://codecov.io/gh/Codex-/await-remote-run) [![GitHub Marketplace](https://img.shields.io/badge/Marketplace-await–remote–run-blue.svg?colorA=24292e&colorB=0366d6&style=flat-square&longCache=true&logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAA4AAAAOCAYAAAAfSC3RAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAM6wAADOsB5dZE0gAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAAERSURBVCiRhZG/SsMxFEZPfsVJ61jbxaF0cRQRcRJ9hlYn30IHN/+9iquDCOIsblIrOjqKgy5aKoJQj4O3EEtbPwhJbr6Te28CmdSKeqzeqr0YbfVIrTBKakvtOl5dtTkK+v4HfA9PEyBFCY9AGVgCBLaBp1jPAyfAJ/AAdIEG0dNAiyP7+K1qIfMdonZic6+WJoBJvQlvuwDqcXadUuqPA1NKAlexbRTAIMvMOCjTbMwl1LtI/6KWJ5Q6rT6Ht1MA58AX8Apcqqt5r2qhrgAXQC3CZ6i1+KMd9TRu3MvA3aH/fFPnBodb6oe6HM8+lYHrGdRXW8M9bMZtPXUji69lmf5Cmamq7quNLFZXD9Rq7v0Bpc1o/tp0fisAAAAASUVORK5CYII=)](https://github.com/marketplace/actions/await-remote-run)

Await the completion of a foreign repository Workflow Run given the Run ID.

This Action exists as a workaround for the issue where you cannot await the completion of a dispatched action.

This action requires being able to get the run ID from a dispatched action, this can be achieved through another Action i've created, [return-dispatch](https://github.com/Codex-/return-dispatch).

Should a remote workflow run fail, this action will attempt to output which step failed, with a link to the workflow run itself.

An example using both of these actions is documented below.

## Usage

Once you have configured your remote repository to work as expected with the `return-dispatch` action, include `await-remote-run` as described below.

```yaml
steps:
  - name: Dispatch an action and get the run ID
    uses: codex-/return-dispatch@v1
    id: return_dispatch
    with:
      token: ${{ github.token }}
      repo: repository-name
      owner: repository-owner
      workflow: automation-test.yml
  - name: Await Run ID ${{ steps.return_dispatch.outputs.run_id }}
    uses: Codex-/await-remote-run@v1.0.0
    with:
      token: ${{ github.token }}
      repo: return-dispatch
      owner: codex-
      run_id: ${{ steps.return_dispatch.outputs.run_id }}
      run_timeout_seconds: 300 # Optional
      poll_interval_ms: 5000 # Optional
```

### Permissions Required

The permissions required for this action to function correctly are:

- `repo` scope
  - You may get away with simply having `repo:public_repo`
  - `repo` is definitely needed if the repository is private.
- `actions:read`

### APIs Used

For the sake of transparency please note that this action uses the following API calls:

- [Get a workflow run](https://docs.github.com/en/rest/reference/actions#get-a-workflow-run)
  - GET `/repos/{owner}/{repo}/actions/runs/{run_id}`
  - Permissions:
    - `repo`
    - `actions:read`
- [List jobs for a workflow run](https://docs.github.com/en/rest/reference/actions#list-jobs-for-a-workflow-run)
  - GET `/repos/{owner}/{repo}/actions/runs/{run_id}/jobs`
  - Permissions:
    - `repo`
    - `actions:read`

For more information please see [api.ts](./src/api.ts).

## Where does this help?

If you want to use the result of a Workflow Run from a remote repository to complete a check locally, i.e. you have automated tests on another repository and don't want the local checks to pass if the remote fails.


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v1.10.0](https://github.com/Codex-/await-remote-run/compare/v1.9.0...v1.10.0)

- chore(deps): lock file maintenance [`#174`](https://github.com/Codex-/await-remote-run/pull/174)
- chore(deps): update dependency release-it to v16 [`#173`](https://github.com/Codex-/await-remote-run/pull/173)
- chore(deps): update all non-major dependencies [`#169`](https://github.com/Codex-/await-remote-run/pull/169)
- chore(deps): update dependency prettier to v3 [`#172`](https://github.com/Codex-/await-remote-run/pull/172)
- build(deps-dev): bump word-wrap from 1.2.3 to 1.2.4 [`#168`](https://github.com/Codex-/await-remote-run/pull/168)
- chore(deps): lock file maintenance [`#167`](https://github.com/Codex-/await-remote-run/pull/167)
- build(deps): bump vite from 4.2.1 to 4.3.9 [`#162`](https://github.com/Codex-/await-remote-run/pull/162)
- build(deps): bump vm2 from 3.9.17 to 3.9.18 [`#160`](https://github.com/Codex-/await-remote-run/pull/160)
- build(deps): bump vm2 from 3.9.16 to 3.9.17 [`#158`](https://github.com/Codex-/await-remote-run/pull/158)
- build(deps): bump vm2 from 3.9.15 to 3.9.16 [`#157`](https://github.com/Codex-/await-remote-run/pull/157)
- build(deps): bump vm2 from 3.9.14 to 3.9.15 [`#156`](https://github.com/Codex-/await-remote-run/pull/156)
- chore(deps): add dependency @typescript-eslint/eslint-plugin [`abe6a73`](https://github.com/Codex-/await-remote-run/commit/abe6a73e5bb6b1ada12463819774b1728815da31)
- test: Use v8 for test coverage [`042e8b7`](https://github.com/Codex-/await-remote-run/commit/042e8b7e6af1fefb794d2cc6a201d8745d00c1de)
- chore(deps): update dependency eslint-plugin-github to v4.10.0 [`fb77a3e`](https://github.com/Codex-/await-remote-run/commit/fb77a3e1d2a8e9da5ef740977b888d759315b751)

#### [v1.9.0](https://github.com/Codex-/await-remote-run/compare/v1.8.0...v1.9.0)

> 2 April 2023

- chore(deps): update dependency typescript to v5 [`#153`](https://github.com/Codex-/await-remote-run/pull/153)
- chore(deps): update all non-major dependencies [`#152`](https://github.com/Codex-/await-remote-run/pull/152)
- chore(deps): update all non-major dependencies [`#151`](https://github.com/Codex-/await-remote-run/pull/151)
- chore(deps): lock file maintenance [`#146`](https://github.com/Codex-/await-remote-run/pull/146)
- build(deps): bump cacheable-request from 10.2.5 to 10.2.7 [`#150`](https://github.com/Codex-/await-remote-run/pull/150)
- chore(deps): update all non-major dependencies [`#149`](https://github.com/Codex-/await-remote-run/pull/149)
- chore(deps): update all non-major dependencies to ^0.28.4 [`c831ffb`](https://github.com/Codex-/await-remote-run/commit/c831ffb9b8e3b38f22dc7d24ab4f96a24c6c7ca7)
- Release 1.9.0 [`ebdef18`](https://github.com/Codex-/await-remote-run/commit/ebdef1805ef1e8b612a70453a048eab968710613)
- chore(deps): update renovatebot config. [`9558cf0`](https://github.com/Codex-/await-remote-run/commit/9558cf02ad4513c9aa432c8b2321eaf830811d50)

#### [v1.8.0](https://github.com/Codex-/await-remote-run/compare/v1.7.0...v1.8.0)

> 1 February 2023

- chore(deps): lock file maintenance [`#145`](https://github.com/Codex-/await-remote-run/pull/145)
- chore(deps): update all non-major dependencies [`#144`](https://github.com/Codex-/await-remote-run/pull/144)
- chore(deps): lock file maintenance [`#143`](https://github.com/Codex-/await-remote-run/pull/143)
- chore(deps): update vitest monorepo to ^0.27.0 [`#139`](https://github.com/Codex-/await-remote-run/pull/139)
- chore(deps): lock file maintenance [`#137`](https://github.com/Codex-/await-remote-run/pull/137)
- chore(deps): update vitest monorepo to ^0.26.0 [`#135`](https://github.com/Codex-/await-remote-run/pull/135)
- Release 1.8.0 [`9b366bb`](https://github.com/Codex-/await-remote-run/commit/9b366bb40027da93b8226ae8eab1fe760483bb15)
- chore(deps): update dependency prettier to v2.8.3 [`40ea9f5`](https://github.com/Codex-/await-remote-run/commit/40ea9f5ecbbef37745b166c9593752c242a2d7ca)
- chore(deps): update dependency prettier to v2.8.2 [`1cd4ccc`](https://github.com/Codex-/await-remote-run/commit/1cd4ccc612d8347cd32cb83db664d7fcdddbb036)

#### [v1.7.0](https://github.com/Codex-/await-remote-run/compare/v1.6.0...v1.7.0)

> 12 December 2022

- chore(deps): lock file maintenance [`#132`](https://github.com/Codex-/await-remote-run/pull/132)
- chore(deps): lock file maintenance [`#130`](https://github.com/Codex-/await-remote-run/pull/130)
- test: convert tests to use vitest. [`c7ac659`](https://github.com/Codex-/await-remote-run/commit/c7ac659c478c2253bac2fc94af24795175118a3b)
- chore(deps): update dependency esbuild to ^0.16.0 [`b7ff644`](https://github.com/Codex-/await-remote-run/commit/b7ff64423f6cfe0ad77e97b9ddd26a7942b4e255)
- chore(deps): bump root dependencies in package. [`2bb9844`](https://github.com/Codex-/await-remote-run/commit/2bb98443ba8d433a2ff1e5d1d72a282dcd6a4dd5)

#### [v1.6.0](https://github.com/Codex-/await-remote-run/compare/v1.5.1...v1.6.0)

> 30 October 2022

- chore(deps): lock file maintenance [`#126`](https://github.com/Codex-/await-remote-run/pull/126)
- chore(deps): lock file maintenance [`#125`](https://github.com/Codex-/await-remote-run/pull/125)
- chore: bump dependencies. [`33dd189`](https://github.com/Codex-/await-remote-run/commit/33dd189fa861b7a9e1615c6a8fe3a396d3fbbb3a)
- Release 1.6.0 [`91f5d9b`](https://github.com/Codex-/await-remote-run/commit/91f5d9b23aafd4992ac391e5c921a354e4e1af15)
- cicd: Use correct ref for PR events. [`14d67e1`](https://github.com/Codex-/await-remote-run/commit/14d67e1e77e1b37c90c89186944e0b6f1969c663)

#### [v1.5.1](https://github.com/Codex-/await-remote-run/compare/v1.5.0...v1.5.1)

> 28 September 2022

- build(deps): bump vm2 from 3.9.10 to 3.9.11 [`#124`](https://github.com/Codex-/await-remote-run/pull/124)

src/action.ts
import * as core from "@actions/core";

const RUN_TIMEOUT_SECONDS = 5 * 60;
const POLL_INTERVAL_MS = 5000;

/**
 * action.yaml definition.
 */
export interface ActionConfig {
  /**
   * GitHub API token for making requests.
   */
  token: string;

  /**
   * Repository of the action to await.
   */
  repo: string;

  /**
   * Owner of the given repository.
   */
  owner: string;

  /**
   * Run ID to await the completion of.
   */
  runId: number;

  /**
   * Time until giving up on the completion of an action.
   * @default 300
   */
  runTimeoutSeconds: number;

  /**
   * Frequency to poll the action for a status.
   * @default 2500
   */
  pollIntervalMs: number;
}

export function getConfig(): ActionConfig {
  return {
    token: core.getInput("token", { required: true }),
    repo: core.getInput("repo", { required: true }),
    owner: core.getInput("owner", { required: true }),
    runId: getRunIdFromValue(core.getInput("run_id")),
    runTimeoutSeconds:
      getNumberFromValue(core.getInput("run_timeout_seconds")) ||
      RUN_TIMEOUT_SECONDS,
    pollIntervalMs:
      getNumberFromValue(core.getInput("poll_interval_ms")) || POLL_INTERVAL_MS,
  };
}

function getRunIdFromValue(value: string): number {
  const id = getNumberFromValue(value);
  if (id === undefined) {
    throw new Error("Run ID must be provided.");
  }
  return id;
}

function getNumberFromValue(value: string): number | undefined {
  if (value === "") {
    return undefined;
  }

  try {
    const num = parseInt(value);

    if (isNaN(num)) {
      throw new Error("Parsed value is NaN");
    }

    return num;
  } catch {
    throw new Error(`Unable to parse value: ${value}`);
  }
}
